hack/verify-license.py

```diff
@@ -0,0 +1,945 @@
+}
+
+# Not all directories have LICENSE files in expected formats. Override them here:
+WEIRD_LICENSE_NAMES = {
+    'vendor/bitbucket.org/ww/goautoneg': 'README.txt',
+}
+
+# Some licenses are too far outside of the normal format to work.
+LICENSES_HASH = {
+    '5b3c09a5f69b76c5ccefe80a2c38293cbbe87b28':
+        ('vendor/bitbucket.org/ww/goautoneg/README.txt', 'BSD (3-Clause)'),
+    'b782e7268198e14de43085d29b562f726155f3f5':
+        ('vendor/github.com/heketi/heketi/LICENSE',
+         'Dual licensed (API under Apache 2.0)'),
+    '087552949856c17dd01c9beeb9b78f04535eaa40':
+        ('vendor/github.com/storageos/go-api/LICENCE',
+         'Dual licensed (MIT/BSD)'),
+    'f1adc08862de0a11fd79ebf42560ebf0c6fc163b':
+        ('vendor/github.com/daviddengcn/go-colortext/LICENSE',
+         'Dual licensed (MIT/BSD)'),
+    '271aeaf56ee621c5accfc2a9db0b10717e038eaf':
+        ('vendor/github.com/ghodss/yaml/LICENSE', 'Dual licensed (MIT/BSD)'),
+    'db170a1a3711f86e4ce228a9fe74e101315900f8':
+        ('vendor/github.com/JeffAshton/win_pdh/LICENSE', 'BSD (3-Clause)'),
+}
+
+# Ignore replacements in these forms.
+# Format is from_string: {target_string}
+IGNORE_SUBS = {
+    '': {
+        'The MIT License (MIT)', 'The MIT License', 'MIT License',
+        'ISC License',
+        '>', '> >'
+    },
+    '1.': {'', '*', '-'},
+    '2.': {'', '*', '-'},
+    '3.': {'', '*'},
+    '(a)': {''}, '(b)': {''}, '(c)': {''}, '(d)': {''},
+    'HOLDER': {'OWNER'},
+    'OWNER': {'HOLDER'},
+    'http://www.apache.org/licenses/': {'https://www.apache.org/licenses/'},
+    'http://www.apache.org/licenses/LICENSE-2.0': {
+        'https://www.apache.org/licenses/LICENSE-2.0'},
+    ('The views and conclusions contained in the software and documentation are'
+     ' those of the authors and should not be interpreted as representing '
+     'official policies, either expressed or implied, of the FreeBSD Project.'):
+     {''}
+}
+
+# These strings are what *should* be replaced in licenses to indicate years and owners.
+COPYRIGHT_NAMES = {
+    '<YEAR> <COPYRIGHT HOLDER>',
+    '<YEAR> <OWNER>',
+    '[yyyy] [name of copyright owner]',
+    'copyright holder',
+    'the copyright holder',
+}
+
+# Ignore punctuation differences.
+PUNCT_RE = re.compile(r'[][(){},.]+')
+
+
+def find_licenses(root):
+    '''
+    Find all license files by walking from a root.
+
+    Args:
+        root: the directory to search for licenses inside
+    Returns:
+        licenses: a list of LICENSE files to validate.
+        uncovered: a list of directories that are missing LICENSE files.
+    '''
+    licenses = []
+    uncovered = []
+    for dirpath, dirnames, filenames in os.walk(root, followlinks=True):
+        if dirpath in WEIRD_LICENSE_NAMES:
+            path = os.path.join(dirpath, WEIRD_LICENSE_NAMES[dirpath])
+            if os.path.exists(path):
+                licenses.append(path)
+                dirnames[:] = []  # don't walk further
```

Comment:
Wow, this is documented `os.walk()` behavior, crazy!

https://docs.python.org/2/library/os.html#os.walk